server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SubAddressing.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.mailet.Mail;
+import org.apache.mailet.StorageDirective;
+import org.apache.mailet.base.GenericMailet;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+
+import static org.apache.james.mailbox.MessageManager.MailboxMetaData.RecentMode.IGNORE;
+
+public class SubAddressing extends GenericMailet {
+    private final UsersRepository usersRepository;
+    private final MailboxManager mailboxManager;
+    private final MetricFactory metricFactory;
+
+    private static final Logger LOG = LoggerFactory.getLogger(SubAddressing.class);
+
+    @Inject
+    public SubAddressing(UsersRepository usersRepository, @Named("mailboxmanager") MailboxManager mailboxManager, MetricFactory metricFactory) {
+        this.usersRepository = usersRepository;
+        this.mailboxManager = mailboxManager;
+        this.metricFactory = metricFactory;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        mail.getRecipients().forEach(recipient ->
+            recipient.getLocalPartDetails(UsersRepository.LOCALPART_DETAIL_DELIMITER)
+                .ifPresent(
+                    Throwing.consumer(targetFolder -> postIfHasRight(mail, recipient, targetFolder))
+                )
+        );
+    }
+
+    private void postIfHasRight(Mail mail, MailAddress recipient, String targetFolder) throws UsersRepositoryException {
+        if(hasPostRight(recipient, targetFolder)) {
+            StorageDirective.builder().targetFolders(ImmutableList.of(targetFolder)).build()
+                .encodeAsAttributes(usersRepository.getUsername(recipient))
+                .forEach(mail::setAttribute);
+        } else {
+            LOG.info("{} tried to address {}'s subfolder `{}` but did not have the right to", mail.getMaybeSender().toString(), recipient, targetFolder);
+        }
+    }
+
+    private boolean hasPostRight(MailAddress recipient, String targetFolder) {
+
+        try {
+            Username recipientUsername = Username.fromLocalPartWithoutDomain(recipient.getLocalPartWithoutDetails(UsersRepository.LOCALPART_DETAIL_DELIMITER));
+            MailboxSession session = mailboxManager.createSystemSession(recipientUsername);
+
+            MailboxACL aclCommand = mailboxManager
+                .getMailbox(MailboxPath.forUser(recipientUsername, targetFolder), session)
+                .getMetaData(IGNORE, session, MessageManager.MailboxMetaData.FetchGroup.NO_COUNT)
+                .getACL();
+
+            return aclCommand.getEntries().containsKey(MailboxACL.ANYBODY_KEY)
+                && aclCommand.getEntries()
+                .get(MailboxACL.ANYBODY_KEY)
+                .equals(MailboxACL.Rfc4314Rights.of(Collections.singletonList(MailboxACL.Right.Post)));
+
+        } catch (MailboxException e) {
+            return false;
```

Comment:
Error handling is hard...

Problem 1: we are silenting potential problems here. It's bad! We likely shall log something when recovering from an exception.

Problem 2: recovering makes sense when facing MailboxNotFoundException -> OK


But mailbox exception is wider than this and includes other class of issue. Would it make sense to also recover from those? Or would ot be better to forward issues for problem we do not understand so that they get later retried?

Reply:
forwarding of course... :)